tsfc/spectral.py

```diff
@@ -44,14 +32,86 @@ def Integrals(expressions, quadrature_multiindex, argument_multiindices, paramet
 
     :returns: list of integral representations
     """
+    # Rewrite: a / b => a * (1 / b)
+    expressions = replace_division(expressions)
```

Comment:
OK, I see.  A little bit of experimentation, using GCC 7.1:

```c
#define VEC 4

void kernel(double * __restrict__ A, double *  C, double *  B, double D)
{
    for (int j = 0; j < 128; j++) {
        for (int k = 0; k < VEC; k++) {
            A[j*VEC + k] += C[j*VEC + k]*B[j*VEC + k] / D;
        }
    }
}
```

If I compile with `-O3 -march=broadwell` I get:

```asm
kernel(double*, double*, double*, double):
  vbroadcastsd ymm0, xmm0
  xor eax, eax
.L2:
  vmovupd ymm1, YMMWORD PTR [rsi+rax]
  vmulpd ymm1, ymm1, YMMWORD PTR [rdx+rax]
  vdivpd ymm1, ymm1, ymm0
  vaddpd ymm1, ymm1, YMMWORD PTR [rdi+rax]
  vmovupd YMMWORD PTR [rdi+rax], ymm1
  add rax, 32
  cmp rax, 4096
  jne .L2
  vzeroupper
  ret
```

So B is broadcast across the vector lanes out of the loop, but then the division happens in the loop.

If I add `-ffast-math` (meaning that the optimisations don't have to maintain IEEE compliance, which, amongst other things allows the compiler to make the substitution `a/b => a*(1/b)` then I get:

```asm
kernel(double*, double*, double*, double):
  vmovsd xmm1, QWORD PTR .LC0[rip]
  xor eax, eax
  vdivsd xmm1, xmm1, xmm0
  vbroadcastsd ymm1, xmm1
.L2:
  vmovupd ymm0, YMMWORD PTR [rsi+rax]
  vmulpd ymm0, ymm0, YMMWORD PTR [rdx+rax]
  vfmadd213pd ymm0, ymm1, YMMWORD PTR [rdi+rax]
  vmovupd YMMWORD PTR [rdi+rax], ymm0
  add rax, 32
  cmp rax, 4096
  jne .L2
  vzeroupper
  ret
.LC0:
  .long 0
  .long 1072693248
```

So we perform the division out of the loop, and inside it's turned into a multiplication.

If I manually replace ` / D => * (1/D)` then I get

```asm
kernel(double*, double*, double*, double):
  vmovsd xmm1, QWORD PTR .LC0[rip]
  xor eax, eax
  vdivsd xmm1, xmm1, xmm0
  vbroadcastsd ymm1, xmm1
.L2:
  vmovupd ymm0, YMMWORD PTR [rsi+rax]
  vmulpd ymm0, ymm0, YMMWORD PTR [rdx+rax]
  vfmadd213pd ymm0, ymm1, YMMWORD PTR [rdi+rax]
  vmovupd YMMWORD PTR [rdi+rax], ymm0
  add rax, 32
  cmp rax, 4096
  jne .L2
  vzeroupper
  ret
.LC0:
  .long 0
  .long 1072693248
```

Which is as if I'd said `-ffast-math`.

We currently don't put `-ffast-math` in the compiler flags, which I think is a reasonable position.